pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ServerCnx.java

```diff
@@ -1307,7 +1326,25 @@ protected void handleProducer(final CommandProducer cmdProducer) {
                         producerId, schema == null ? "absent" : "present");
             }
 
-            service.getOrCreateTopic(topicName.toString()).thenCompose((Topic topic) -> {
+            boolean createTopicIfDoesNotExist = isAuthorizedToCreateTopic
+                    && service.isAllowAutoTopicCreation(topicName.toString());
+            service.getTopic(topicName.toString(), createTopicIfDoesNotExist).thenCompose(optTopic -> {
+                if (optTopic.isEmpty()) {
+                    if (isAuthorizedToCreateTopic) {
+                        return FutureUtil
+                                .failedFuture(new TopicNotFoundException(
+                                        "Topic " + topicName + " does not exist"));
+                    } else {
+                        String msg = "Topic to produce does not exists and the Client is not"
+                                + " authorized to create topic";
+                        log.warn("[{}] {} with role {}", remoteAddress, msg, getPrincipal());
+                        ctx.writeAndFlush(Commands.newError(requestId, ServerError.AuthorizationError,
+                                msg));
+                        return null;
```

Comment:
Similarly, we need to throw an exception here. The error handling includes handling the dangling `producer` future, and we need to do that to prevent a build up of unnecessary futures.

Reply:
See previous thread.